File: scanCodebaseCoverage.m
function result = scanCodebaseCoverage(codebaseDir, callSequence, functionTable)
% SCANCODBASECOVERAGE Compare profiler call data against a codebase directory.
%
%   result = scanCodebaseCoverage(codebaseDir, callSequence, functionTable)
%
%   Inputs:
%     codebaseDir   - Root directory of the codebase to scan
%     callSequence  - struct array from parseProfilerPhases (can be empty [])
%     functionTable - FunctionTable from the profiler data
%
%   Output:
%     result - struct with fields:
%       .codebaseFunctions - cell array of all .m function names found
%       .codebaseFiles     - cell array of full file paths
%       .calledFunctions   - cell array of functions that appear in the profiler data
%       .uncalledFunctions - cell array of functions not found in profiler data
%       .uncalledFiles     - cell array of file paths for uncalled functions
%       .coveragePercent   - percentage of codebase functions that were called
%       .coverageTable     - table with columns: Name, File, Called, CallCount

    if ~isfolder(codebaseDir)
        error('scanCodebaseCoverage:badDir', ...
            'Directory does not exist: %s', codebaseDir);
    end

    % Scan for all .m files (top-level functions only)
    mFiles = dir(fullfile(codebaseDir, '**', '*.m'));

    % Filter to actual files (not directories)
    mFiles = mFiles(~[mFiles.isdir]);

    nFiles = numel(mFiles);
    codebaseFuncs = cell(nFiles, 1);
    codebaseFiles = cell(nFiles, 1);

    for k = 1:nFiles
        [~, fname, ~] = fileparts(mFiles(k).name);
        codebaseFuncs{k} = fname;
        codebaseFiles{k} = fullfile(mFiles(k).folder, mFiles(k).name);
    end

    % Build the set of functions that appear in profiler data
    % Use both FunctionTable (all profiled functions) and callSequence (phase-specific)
    profilerFuncNames = {};

    % From FunctionTable: extract base function names
    if ~isempty(functionTable)
        for k = 1:numel(functionTable)
            rawName = functionTable(k).FunctionName;
            % FunctionName may be like 'path>funcName' or 'funcName'
            parts = strsplit(rawName, '>');
            baseName = parts{end};
            % Strip any subfunc markers
            dotParts = strsplit(baseName, '.');
            baseName = dotParts{1};
            profilerFuncNames{end+1} = baseName; %#ok<AGROW>
        end
    end

    % From callSequence: extract function names for phase-specific coverage
    phaseCalledNames = {};
    phaseCallCounts = containers.Map();
    if ~isempty(callSequence)
        for k = 1:numel(callSequence)
            if strcmp(callSequence(k).event, 'enter')
                rawName = callSequence(k).funcName;
                parts = strsplit(rawName, '>');
                baseName = parts{end};
                dotParts = strsplit(baseName, '.');
                baseName = dotParts{1};
                phaseCalledNames{end+1} = baseName; %#ok<AGROW>
                if phaseCallCounts.isKey(baseName)
                    phaseCallCounts(baseName) = phaseCallCounts(baseName) + 1;
                else
                    phaseCallCounts(baseName) = 1;
                end
            end
        end
    end

    profilerFuncNames = unique(profilerFuncNames);
    phaseCalledNames  = unique(phaseCalledNames);

    % Compare against codebase
    calledMask = false(nFiles, 1);
    callCounts = zeros(nFiles, 1);

    for k = 1:nFiles
        fname = codebaseFuncs{k};
        % Check against full profiler data
        if any(strcmpi(profilerFuncNames, fname))
            calledMask(k) = true;
        end
        % Check phase-specific call count
        if phaseCallCounts.isKey(fname)
            callCounts(k) = phaseCallCounts(fname);
        elseif any(strcmpi(phaseCalledNames, fname))
            callCounts(k) = 1;
        end
    end

    calledFuncs   = codebaseFuncs(calledMask);
    uncalledFuncs = codebaseFuncs(~calledMask);
    uncalledFiles = codebaseFiles(~calledMask);

    if nFiles > 0
        coveragePct = 100 * sum(calledMask) / nFiles;
    else
        coveragePct = 0;
    end

    % Build a summary table
    calledStr = cell(nFiles, 1);
    for k = 1:nFiles
        if calledMask(k)
            calledStr{k} = 'Yes';
        else
            calledStr{k} = 'No';
        end
    end

    covTable = table(codebaseFuncs, codebaseFiles, calledStr, callCounts, ...
        'VariableNames', {'Name', 'File', 'Called', 'CallCount'});

    % Sort: uncalled first, then by name
    covTable = sortrows(covTable, {'Called', 'Name'}, {'ascend', 'ascend'});

    % Package output
    result.codebaseFunctions = codebaseFuncs;
    result.codebaseFiles     = codebaseFiles;
    result.calledFunctions   = calledFuncs;
    result.uncalledFunctions = uncalledFuncs;
    result.uncalledFiles     = uncalledFiles;
    result.coveragePercent   = coveragePct;
    result.coverageTable     = covTable;
end

File: Python_Version/scan_codebase_coverage.py
"""Compare profiler call data against a codebase directory.

Scans a directory tree for .m files and cross-references them with
profiler data to determine which functions were called and which were not.
"""

from dataclasses import dataclass
from pathlib import Path

from .parse_profiler_phases import CallEvent


@dataclass
class CoverageEntry:
    """Coverage status for a single codebase function."""
    name: str
    file: str
    called: bool
    call_count: int


@dataclass
class CoverageResult:
    """Result of codebase coverage analysis."""
    codebase_functions: list[str]
    codebase_files: list[str]
    called_functions: list[str]
    uncalled_functions: list[str]
    uncalled_files: list[str]
    coverage_percent: float
    entries: list[CoverageEntry]


def scan_codebase_coverage(
    codebase_dir: str | Path,
    call_sequence: list[CallEvent],
    function_table: list[dict],
) -> CoverageResult:
    """Compare profiler data against a codebase directory.

    Args:
        codebase_dir: Root directory to scan for .m files.
        call_sequence: Call sequence from parse_profiler_phases (can be empty).
        function_table: FunctionTable from the profiler data.

    Returns:
        CoverageResult with per-function coverage information.
    """
    codebase_dir = Path(codebase_dir)
    if not codebase_dir.is_dir():
        raise NotADirectoryError(f"Directory does not exist: {codebase_dir}")

    # Scan for all .m files
    m_files = sorted(codebase_dir.rglob("*.m"))

    codebase_funcs: list[str] = []
    codebase_files: list[str] = []
    for f in m_files:
        if f.is_file():
            codebase_funcs.append(f.stem)
            codebase_files.append(str(f))

    n_files = len(codebase_funcs)

    # Build set of functions from FunctionTable (all profiled functions)
    profiler_func_names: set[str] = set()
    if function_table:
        for entry in function_table:
            raw_name = str(entry.get("FunctionName", ""))
            # FunctionName may be like "path>funcName"
            parts = raw_name.split(">")
            base_name = parts[-1].split(".")[0]
            if base_name:
                profiler_func_names.add(base_name.lower())

    # Build call counts from call_sequence (phase-specific)
    phase_call_counts: dict[str, int] = {}
    if call_sequence:
        for ev in call_sequence:
            if ev.event == "enter":
                raw_name = ev.func_name
                parts = raw_name.split(">")
                base_name = parts[-1].split(".")[0]
                base_lower = base_name.lower()
                phase_call_counts[base_lower] = phase_call_counts.get(base_lower, 0) + 1

    # Compare
    called_mask = [False] * n_files
    call_counts = [0] * n_files

    for i, fname in enumerate(codebase_funcs):
        fname_lower = fname.lower()
        if fname_lower in profiler_func_names:
            called_mask[i] = True
        if fname_lower in phase_call_counts:
            call_counts[i] = phase_call_counts[fname_lower]

    called_funcs = [f for f, m in zip(codebase_funcs, called_mask) if m]
    uncalled_funcs = [f for f, m in zip(codebase_funcs, called_mask) if not m]
    uncalled_files_list = [
        f for f, m in zip(codebase_files, called_mask) if not m
    ]

    coverage_pct = (sum(called_mask) / n_files * 100) if n_files > 0 else 0.0

    entries = [
        CoverageEntry(
            name=codebase_funcs[i],
            file=codebase_files[i],
            called=called_mask[i],
            call_count=call_counts[i],
        )
        for i in range(n_files)
    ]
    # Sort: uncalled first, then by name
    entries.sort(key=lambda e: (e.called, e.name.lower()))

    return CoverageResult(
        codebase_functions=codebase_funcs,
        codebase_files=codebase_files,
        called_functions=called_funcs,
        uncalled_functions=uncalled_funcs,
        uncalled_files=uncalled_files_list,
        coverage_percent=coverage_pct,
        entries=entries,
    )
